versioned/transfer/src/main/java/org/projectnessie/versioned/transfer/ExportPersist.java

```diff
@@ -105,21 +105,34 @@ private HeadsAndForkPoints scanAllReferences(Consumer<CommitObj> commitHandler)
     referenceLogic
         .queryReferences(referencesQuery())
         .forEachRemaining(
-            ref -> {
-              for (Iterator<CommitObj> commitIter =
-                      commitLogic.commitLog(commitLogQuery(ref.pointer()));
-                  commitIter.hasNext(); ) {
-                CommitObj commit = commitIter.next();
-                if (!identify.handleCommit(commit)) {
-                  break;
-                }
-                commitHandler.accept(commit);
-              }
-            });
+            ref ->
+                scanCommitLogChain(
+                    commitLogic.commitLog(commitLogQuery(ref.pointer())),
+                    identify,
+                    commitHandler,
+                    commitLogic));
 
     return identify.finish();
   }
 
+  private void scanCommitLogChain(
+      Iterator<CommitObj> commitIter,
+      IdentifyHeadsAndForkPoints identify,
+      Consumer<CommitObj> commitHandler,
+      CommitLogic commitLogic) {
+    while (commitIter.hasNext()) {
+      CommitObj commit = commitIter.next();
+      if (!identify.handleCommit(commit)) {
+        break;
+      }
+      commitHandler.accept(commit);
+      for (ObjId objId : commit.secondaryParents()) {
+        scanCommitLogChain(
```

Comment:
Can we avoid recursion here? I suppose the commit DAG can have a lot of nested branch points.

Reply:
Hmm how? We need to scan the whole lineage of each secondary parent. (Note that the scan stops as soon as a commit has been seen before, so we are not re-scanning commits more than once.)

Reply:
We could still have a cascade of merge parents that is longer than the java stack, I believe... we'd have to recurse into them even if we do not really traverse any father :thinking:

Reply:
You are right, let's remove recursion.